src/libraries/System.Text.Json/src/System/Collections/Generic/ReferenceEqualityComparer.netstandard.cs

```diff
@@ -0,0 +1,23 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Runtime.CompilerServices;
+
+namespace System.Collections.Generic
+{
+    /// <summary>
+    /// Passed to the <see cref="DefaultReferenceResolver._objectToReferenceIdMap"/> meant for serialization.
+    /// It forces the dictionary to do a ReferenceEquals comparison when comparing the TKey object.
+    /// </summary>
+    internal sealed class ReferenceEqualityComparer : IEqualityComparer<object>, IEqualityComparer
```

Comment:
Can we directly source include `...System/Collections/Generic/ReferenceEqualityComparer.cs` into this project rather than having a duplicate? Makes it easier to keep in sync (rather than having duplicate copies), just in case changes are made to one that should have been made to both.

Reply:
Would introducing a new public type in the _src_ but not the _ref_ potentially affect consumers of your package?

Reply:
#ifdef and mark it internal when OOB for similar to what we did for ArrayBufferWriter?
https://github.com/dotnet/runtime/blob/7e128196c0b477344888042c7c508f4624dc65ef/src/libraries/Common/src/System/Buffers/ArrayBufferWriter.cs#L12-L17